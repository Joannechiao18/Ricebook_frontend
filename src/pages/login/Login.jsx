import React, { useState, useContext } from "react";
import { Link, useNavigate } from "react-router-dom";

import "./login.scss";
import "bootstrap/dist/css/bootstrap.min.css";
import styled from "styled-components";
import { AuthContext } from "../../context/authContext";
import DatePicker from "react-datepicker";
import { useDispatch } from "react-redux";
import { login } from "../../actions/authActions"; // Assuming you've stored it in an 'actions' folder
import GoogleIcon from "@mui/icons-material/Google";
// Styled Components for the buttons
const LoginButton = styled.button`
  background-color: #938eef;
  color: white;
  border: none;
  padding: 10px 15px;
  border-radius: 5px;
  font-size: 15px;
  cursor: pointer;
  transition: background-color 0.3s ease;
  font-weight: 600;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto,
    "Helvetica Neue", Arial, "Noto Sans", sans-serif, "Apple Color Emoji",
    "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji";

  &:hover {
    background-color: #7a75d6;
    animation: pulse 0.6s infinite;
  }
`;

const RegisterButton = styled(LoginButton)`
  background-color: #e3e0e0;
  color: black;

  &:hover {
    background-color: #dedada;
  }
`;

const ThirdPartyLoginButton = styled(LoginButton)`
  background-color: #3b5998; // Adjust the color as needed
  margin-right: 20px; // Increase the margin for a larger gap

  &:hover {
    background-color: #324c85; // Adjust the hover color as needed
  }
`;

const Login = () => {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [loginError, setLoginError] = useState(""); // State to store login error message
  const navigate = useNavigate();
  const dispatch = useDispatch();

  const handleThirdPartyLogin = () => {
    window.location.href =
      "https://ricebookserveryw187-8fbcb305db50.herokuapp.com/auth/google/callback";
  };

  const handleSubmit = async (e) => {
    e.preventDefault();

    // Check for username and password only if the standard login button was clicked
    if (
      e.nativeEvent.submitter ===
        document.getElementById("standardLoginButton") &&
      (!username || !password)
    ) {
      setLoginError("Username and password are required.");
      return;
    }
    try {
      const loginResponse = await fetch(
        "https://ricebookserveryw187-8fbcb305db50.herokuapp.com/login",
        {
          method: "POST",

          headers: {
            "Content-Type": "application/json",
          },
          body: JSON.stringify({
            username: username,
            password: password,
          }),
          credentials: "include",
        }
      );

      const loginData = await loginResponse.json();
      if (loginResponse.ok) {
        // Fetch avatar
        const avatarResponse = await fetch(
          `https://ricebookserveryw187-8fbcb305db50.herokuapp.com/avatar/${username}`,
          {
            method: "GET",
            credentials: "include",
            headers: {
              "Content-Type": "application/json",
            },
          }
        );

        let avatarUrl = null;
        if (avatarResponse.ok) {
          const avatarData = await avatarResponse.json();
          avatarUrl = avatarData.avatar;
        }

        // Update Redux store with user data and avatar
        dispatch(login({ ...loginData, avatar: avatarUrl })); // Assuming your login action can handle this data structure
        navigate("/");
      } else {
        console.error("Login error:", loginData.error);
        setLoginError(loginData.error || "Invalid login credentials");
      }
    } catch (error) {
      console.error("Network error:", error);
      setLoginError("Network error. Please try again later.");
    }
  };

  return (
    <div className="register d-flex align-items-center vh-100">
      <div className="card mx-auto" style={{ maxWidth: "800px" }}>
        <div className="row g-0">
          <div className="col-md-6 d-flex flex-column justify-content-center align-items-center p-5 text-white bg-primary">
            <h2 className="display-1 mb-4">Hello World.</h2>
            <span className="mb-3">First time?</span>
            <Link to="/register">
              <RegisterButton>Register</RegisterButton>
            </Link>
          </div>
          <div className="col-md-6 d-flex flex-column justify-content-center p-5">
            <h2 className="mb-4">Login</h2>
            {loginError && (
              <div className="alert alert-danger" role="alert">
                {loginError}
              </div>
            )}
            <form onSubmit={handleSubmit}>
              <div className="mb-3">
                <input
                  type="text"
                  className="form-control w-100"
                  placeholder="Username"
                  value={username}
                  onChange={(e) => setUsername(e.target.value)}
                />
              </div>
              <div className="mb-3">
                <input
                  type="password"
                  className="form-control w-100"
                  placeholder="Password"
                  value={password}
                  onChange={(e) => setPassword(e.target.value)}
                />
              </div>
              <div className="d-flex justify-content-between align-items-center">
                <div>
                  <ThirdPartyLoginButton
                    type="button"
                    onClick={handleThirdPartyLogin}
                  >
                    <GoogleIcon>&#xe8d4;</GoogleIcon> Login with Google
                  </ThirdPartyLoginButton>
                </div>
                <div>
                  <LoginButton id="standardLoginButton" type="submit">
                    Login
                  </LoginButton>
                </div>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  );
};

export default Login;
