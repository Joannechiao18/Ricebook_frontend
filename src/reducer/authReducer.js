// authReducer.js
import { createReducer } from "@reduxjs/toolkit";
import { login, register, logout } from "../actions/authActions";

const initialState = {
  currentUser: null,
  isLoggedIn: false,
  profilePic: null,
};

const authReducer = createReducer(initialState, {
  [login]: (state, action) => {
    state.currentUser = action.payload;
    state.isLoggedIn = true;
    state.profilePic = action.payload.profilePic;
  },
  [register]: (state, action) => {
    state.currentUser = action.payload;
    state.isLoggedIn = true;
    state.profilePic = action.payload.profilePic;
  },
  [logout]: (state) => {
    localStorage.removeItem("persist:root");
    localStorage.removeItem("user");
    localStorage.removeItem("userHeadline");
    state.currentUser = null;
    state.isLoggedIn = false;
  },
});

export default authReducer;

export const selectUser = (state) => state.auth.currentUser;
export const selectProfilePic = (state) => state.auth.profilePic;
