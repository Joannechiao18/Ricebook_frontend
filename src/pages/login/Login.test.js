import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import { createStore } from "redux";
import Login from "../login/Login";
import authReducer from "../../reducer/authReducer";
import { BrowserRouter as Router } from "react-router-dom";
import fetchMock from "jest-fetch-mock";

// Mock the useNavigate hook from react-router-dom
const mockNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate,
}));

// Mock the useDispatch hook from react-redux
const mockDispatch = jest.fn();
jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useDispatch: () => mockDispatch,
}));

// Mock global alert
global.alert = jest.fn();

describe("Login Component", () => {
  beforeEach(() => {
    fetchMock.resetMocks();
    mockNavigate.mockClear();
    mockDispatch.mockClear();
    fetchMock.mockClear();
    global.alert.mockClear();

    // Mock the fetch response
    fetchMock.mockResponseOnce(
      JSON.stringify([
        {
          id: 1,
          username: "testuser",
          email: "test@example.com",
          phone: "1234567890",
          address: {
            street: "password",
            zipcode: "12345",
          },
        },
      ])
    );
  });

  it("should log in a previously registered user and navigate to root", async () => {
    // Set up initial state for the Redux store
    const initialState = {
      auth: {
        currentUser: null,
        isLoggedIn: false,
        profilePic: null,
      },
    };
    const store = createStore(authReducer, initialState);

    const { getByPlaceholderText, getByText } = render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );

    // Fill in login details
    fireEvent.change(getByPlaceholderText("Username"), {
      target: { value: "testuser" },
    });
    fireEvent.change(getByPlaceholderText("Password (Street Name)"), {
      target: { value: "password" },
    });

    // Submit the form
    fireEvent.click(getByText("LOGIN"));

    await new Promise((resolve) => setTimeout(resolve, 0));

    // Assertions
    expect(fetchMock).toHaveBeenCalledTimes(1);
    expect(fetchMock).toHaveBeenCalledWith(
      "https://jsonplaceholder.typicode.com/users?username=testuser"
    );
    expect(mockDispatch).toHaveBeenCalled();
    expect(mockNavigate).toHaveBeenCalledWith("/");
  });

  it("should not log in an invalid user", async () => {
    // Set up initial state for the Redux store
    const initialState = {
      auth: {
        currentUser: null,
        isLoggedIn: false,
        profilePic: null,
      },
    };
    const store = createStore(authReducer, initialState);

    const { getByPlaceholderText, getByText } = render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );

    // Mock a failed fetch response
    fetchMock.mockResponseOnce(
      JSON.stringify([]), // Assume no user is found with the provided credentials
      { status: 404, statusText: "Not Found" }
    );

    // Fill in login details
    fireEvent.change(getByPlaceholderText("Username"), {
      target: { value: "invaliduser" },
    });
    fireEvent.change(getByPlaceholderText("Password (Street Name)"), {
      target: { value: "invalidpassword" },
    });

    // Submit the form
    fireEvent.click(getByText("LOGIN"));

    await new Promise((resolve) => setTimeout(resolve, 0));

    // Assertions
    expect(fetchMock).toHaveBeenCalledTimes(1);
    expect(fetchMock).toHaveBeenCalledWith(
      "https://jsonplaceholder.typicode.com/users?username=invaliduser"
    );
    expect(mockDispatch).not.toHaveBeenCalled();
    expect(mockNavigate).not.toHaveBeenCalledWith("/");
    expect(global.alert).toHaveBeenCalledWith("Incorrect username or password");
  });

  it("matches snapshot", () => {
    const initialState = {
      auth: {
        currentUser: null,
        isLoggedIn: false,
        profilePic: null,
      },
    };
    const store = createStore(authReducer, initialState);

    const tree = render(
      <Provider store={store}>
        <Router>
          <Login />
        </Router>
      </Provider>
    );

    expect(tree).toMatchSnapshot();
  });

  // ... other tests ...
});
