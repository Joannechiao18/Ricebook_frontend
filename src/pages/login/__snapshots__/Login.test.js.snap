// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="register d-flex align-items-center vh-100"
      >
        <div
          class="card mx-auto"
          style="max-width: 800px;"
        >
          <div
            class="row g-0"
          >
            <div
              class="col-md-6 d-flex flex-column justify-content-center align-items-center p-5 text-white bg-primary"
            >
              <h2
                class="display-1 mb-4"
              >
                Hello World.
              </h2>
              <span
                class="mb-3"
              >
                First time?
              </span>
              <a
                href="/register"
              >
                <button
                  class="sc-beyTiQ sc-guDMob lfzEJE dKAXvw"
                >
                  REGISTER
                </button>
              </a>
            </div>
            <div
              class="col-md-6 d-flex flex-column justify-content-center p-5"
            >
              <h2
                class="mb-4"
              >
                Login
              </h2>
              <form>
                <div
                  class="mb-3"
                >
                  <input
                    class="form-control w-100"
                    placeholder="Username"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="mb-3"
                >
                  <input
                    class="form-control w-100"
                    placeholder="Password (Street Name)"
                    type="password"
                    value=""
                  />
                </div>
                <div
                  class="d-flex justify-content-center"
                >
                  <button
                    class="sc-beyTiQ lfzEJE"
                    type="submit"
                  >
                    LOGIN
                  </button>
                </div>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="register d-flex align-items-center vh-100"
    >
      <div
        class="card mx-auto"
        style="max-width: 800px;"
      >
        <div
          class="row g-0"
        >
          <div
            class="col-md-6 d-flex flex-column justify-content-center align-items-center p-5 text-white bg-primary"
          >
            <h2
              class="display-1 mb-4"
            >
              Hello World.
            </h2>
            <span
              class="mb-3"
            >
              First time?
            </span>
            <a
              href="/register"
            >
              <button
                class="sc-beyTiQ sc-guDMob lfzEJE dKAXvw"
              >
                REGISTER
              </button>
            </a>
          </div>
          <div
            class="col-md-6 d-flex flex-column justify-content-center p-5"
          >
            <h2
              class="mb-4"
            >
              Login
            </h2>
            <form>
              <div
                class="mb-3"
              >
                <input
                  class="form-control w-100"
                  placeholder="Username"
                  type="text"
                  value=""
                />
              </div>
              <div
                class="mb-3"
              >
                <input
                  class="form-control w-100"
                  placeholder="Password (Street Name)"
                  type="password"
                  value=""
                />
              </div>
              <div
                class="d-flex justify-content-center"
              >
                <button
                  class="sc-beyTiQ lfzEJE"
                  type="submit"
                >
                  LOGIN
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
